### DISCUSSION TITLE
Python file works in Jupyter Notebook but not after embedded in pyinstaller

### DISCUSSION THREAD

**Discussion by jou-hui** [pyinstaller/pyinstaller#5543] [Help] [answered]
Please help. 

Build Log File 
C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan>pyinstaller --onefile Exhaust_fan.py
461 INFO: PyInstaller: 4.2
462 INFO: Python: 3.7.4
463 INFO: Platform: Windows-10-10.0.17763-SP0
469 INFO: wrote C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\Exhaust_fan.spec
486 INFO: UPX is not available.
489 INFO: Extending PYTHONPATH with paths
['C:\\Users\\lim.jouhui\\Desktop\\DRMSA\\Exploratory analysis\\Electrical '
 'Component Sizing\\Exhaust_fan',
 'C:\\Users\\lim.jouhui\\Desktop\\DRMSA\\Exploratory analysis\\Electrical '
 'Component Sizing\\Exhaust_fan']
558 INFO: checking Analysis
559 INFO: Building Analysis because Analysis-00.toc is non existent
560 INFO: Initializing module dependency graph...
565 INFO: Caching module graph hooks...
576 WARNING: Several hooks defined for module 'win32ctypes.core'. Please take care they do not conflict.
592 INFO: Analyzing base_library.zip ...
4182 INFO: Caching module dependency graph...
4309 INFO: running Analysis Analysis-00.toc
4344 INFO: Adding Microsoft.Windows.Common-Controls to dependent assemblies of final executable
  required by c:\users\lim.jouhui\appdata\local\programs\python\python37\python.exe
4429 INFO: Analyzing C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\Exhaust_fan.py
6028 INFO: Processing pre-find module path hook distutils from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-distutils.py'.
6028 INFO: distutils: retargeting to non-venv dir 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib'
7705 INFO: Processing pre-find module path hook site from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-site.py'.
7714 INFO: site: retargeting to fake-dir 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\fake-modules'
11077 INFO: Processing pre-safe import module hook setuptools.extern.six.moves from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-setuptools.extern.six.moves.py'.
17424 INFO: Processing pre-safe import module hook six.moves from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\pre_safe_import_module\\hook-six.moves.py'.
31784 INFO: Processing module hooks...
31784 INFO: Loading module hook 'hook-openpyxl.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
32059 INFO: Loading module hook 'hook-pycparser.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
32063 INFO: Loading module hook 'hook-win32ctypes.core.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
32496 INFO: Loading module hook 'hook-difflib.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32507 INFO: Excluding import of doctest from module difflib
32508 INFO: Loading module hook 'hook-distutils.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32520 INFO: Loading module hook 'hook-distutils.util.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32532 INFO: Excluding import of lib2to3.refactor from module distutils.util
32532 INFO: Loading module hook 'hook-encodings.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32729 INFO: Loading module hook 'hook-heapq.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32741 INFO: Excluding import of doctest from module heapq
32742 INFO: Loading module hook 'hook-importlib_metadata.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32744 INFO: Loading module hook 'hook-lib2to3.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
32890 INFO: Loading module hook 'hook-matplotlib.backends.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
34035 INFO:   Matplotlib backend "GTK3Agg": ignored
    cairo backend requires that pycairo>=1.11.0 or cairocffiis installed
34511 INFO:   Matplotlib backend "GTK3Cairo": ignored
    cairo backend requires that pycairo>=1.11.0 or cairocffiis installed
35168 INFO:   Matplotlib backend "MacOSX": ignored
    cannot import name '_macosx' from 'matplotlib.backends' (c:\users\lim.jouhui\appdata\local\programs\python\python37\lib\site-packages\matplotlib\backends\__init__.py)
35642 INFO:   Matplotlib backend "nbAgg": ignored
    No module named 'IPython'
36136 INFO:   Matplotlib backend "Qt4Agg": ignored
    Failed to import any qt binding
36591 INFO:   Matplotlib backend "Qt4Cairo": ignored
    cairo backend requires that pycairo>=1.11.0 or cairocffiis installed
37070 INFO:   Matplotlib backend "Qt5Agg": ignored
    Failed to import any qt binding
37543 INFO:   Matplotlib backend "Qt5Cairo": ignored
    cairo backend requires that pycairo>=1.11.0 or cairocffiis installed
38211 INFO:   Matplotlib backend "TkAgg": added
38842 INFO:   Matplotlib backend "TkCairo": ignored
    cairo backend requires that pycairo>=1.11.0 or cairocffiis installed
39308 INFO:   Matplotlib backend "WebAgg": ignored
    Traceback (most recent call last):
  File "c:\users\lim.jouhui\appdata\local\programs\python\python37\lib\site-packages\matplotlib\backends\backend_webagg.py", line 26, in <module>
    import tornado
ModuleNotFoundError: No module named 'tornado'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 12, in <module>
  File "c:\users\lim.jouhui\appdata\local\programs\python\python37\lib\site-packages\matplotlib\backends\backend_webagg.py", line 28, in <module>
    raise RuntimeError("The WebAgg backend requires Tornado.")
RuntimeError: The WebAgg backend requires Tornado.
40005 INFO:   Matplotlib backend "WX": ignored
    No module named 'wx'
40552 INFO:   Matplotlib backend "WXAgg": ignored
    No module named 'wx'
40995 INFO:   Matplotlib backend "WXCairo": ignored
    No module named 'wx'
41516 INFO:   Matplotlib backend "agg": added
41981 INFO:   Matplotlib backend "cairo": ignored
    cairo backend requires that pycairo>=1.11.0 or cairocffiis installed
42593 INFO:   Matplotlib backend "pdf": added
43172 INFO:   Matplotlib backend "pgf": added
43644 INFO:   Matplotlib backend "ps": added
44106 INFO:   Matplotlib backend "svg": added
44677 INFO:   Matplotlib backend "template": added
44874 INFO: Loading module hook 'hook-matplotlib.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
45327 INFO: Loading module hook 'hook-multiprocessing.util.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
45327 INFO: Import to be excluded not found: 'test'
45327 INFO: Loading module hook 'hook-numpy.core.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
45448 INFO: Loading module hook 'hook-numpy.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
45457 INFO: Loading module hook 'hook-pandas.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
48297 INFO: Loading module hook 'hook-pickle.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
48310 INFO: Excluding import of argparse from module pickle
48310 INFO: Loading module hook 'hook-pkg_resources.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49096 INFO: Processing pre-safe import module hook win32com from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\pre_safe_import_module\\hook-win32com.py'.
49441 WARNING: Hidden import "pkg_resources.py2_warn" not found!
49442 WARNING: Hidden import "pkg_resources.markers" not found!
49447 INFO: Excluding import of __main__ from module pkg_resources
49448 INFO: Loading module hook 'hook-pytz.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49813 INFO: Loading module hook 'hook-scipy.linalg.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49828 INFO: Loading module hook 'hook-scipy.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49852 INFO: Loading module hook 'hook-scipy.sparse.csgraph.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49864 INFO: Loading module hook 'hook-scipy.spatial.transform.rotation.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49875 INFO: Loading module hook 'hook-scipy.special._ellip_harm_2.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49884 INFO: Loading module hook 'hook-scipy.special._ufuncs.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49894 INFO: Loading module hook 'hook-scipy.stats._stats.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
49903 INFO: Loading module hook 'hook-setuptools.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
50727 INFO: Excluding import of setuptools.py33compat from module setuptools.depends
50727 INFO: Excluding import of setuptools.py33compat from module setuptools.package_index
50727 INFO: Excluding import of setuptools.py27compat from module setuptools.command.easy_install
50727 INFO: Excluding import of setuptools.py27compat from module setuptools.package_index
50727 INFO: Loading module hook 'hook-sqlite3.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
50953 INFO: Loading module hook 'hook-sysconfig.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
50962 INFO: Loading module hook 'hook-xml.dom.domreg.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
50970 INFO: Loading module hook 'hook-xml.etree.cElementTree.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
50996 INFO: Loading module hook 'hook-xml.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
51009 INFO: Loading module hook 'hook-_tkinter.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
51442 INFO: checking Tree
51443 INFO: Building Tree because Tree-00.toc is non existent
51444 INFO: Building Tree Tree-00.toc
51635 INFO: checking Tree
51635 INFO: Building Tree because Tree-01.toc is non existent
51636 INFO: Building Tree Tree-01.toc
51877 INFO: checking Tree
51878 INFO: Building Tree because Tree-02.toc is non existent
51878 INFO: Building Tree Tree-02.toc
51891 INFO: Loading module hook 'hook-pythoncom.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
52443 INFO: Loading module hook 'hook-pywintypes.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
53013 INFO: Loading module hook 'hook-win32com.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\_pyinstaller_hooks_contrib\\hooks\\stdhooks'...
53652 INFO: Loading module hook 'hook-setuptools.msvc.py' from 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks'...
53672 INFO: Excluding import of numpy from module setuptools.msvc
53815 INFO: Looking for ctypes DLLs
53865 INFO: Analyzing run-time hooks ...
53877 INFO: Including run-time hook 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_pkgres.py'
53889 INFO: Including run-time hook 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_win32api.py'
53900 INFO: Including run-time hook 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_win32comgenpy.py'
53913 INFO: Including run-time hook 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_multiprocessing.py'
53929 INFO: Including run-time hook 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_mplconfig.py'
53941 INFO: Including run-time hook 'c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth__tkinter.py'
53981 INFO: Looking for dynamic libraries
56759 INFO: Looking for eggs
56759 INFO: Using Python library c:\users\lim.jouhui\appdata\local\programs\python\python37\python37.dll
56760 INFO: Found binding redirects:
[]
56783 INFO: Warnings written to C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\build\Exhaust_fan\warn-Exhaust_fan.txt
57014 INFO: Graph cross-reference written to C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\build\Exhaust_fan\xref-Exhaust_fan.html
57166 INFO: checking PYZ
57166 INFO: Building PYZ because PYZ-00.toc is non existent
57167 INFO: Building PYZ (ZlibArchive) C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\build\Exhaust_fan\PYZ-00.pyz
60311 INFO: Building PYZ (ZlibArchive) C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\build\Exhaust_fan\PYZ-00.pyz completed successfully.
60363 INFO: checking PKG
60363 INFO: Building PKG because PKG-00.toc is non existent
60364 INFO: Building PKG (CArchive) PKG-00.pkg
85120 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
85180 INFO: Bootloader c:\users\lim.jouhui\appdata\local\programs\python\python37\lib\site-packages\PyInstaller\bootloader\Windows-64bit\run.exe
85180 INFO: checking EXE
85181 INFO: Building EXE because EXE-00.toc is non existent
85181 INFO: Building EXE from EXE-00.toc
85232 INFO: Copying icons from ['c:\\users\\lim.jouhui\\appdata\\local\\programs\\python\\python37\\lib\\site-packages\\PyInstaller\\bootloader\\images\\icon-console.ico']
85246 INFO: Writing RT_GROUP_ICON 0 resource with 104 bytes
85246 INFO: Writing RT_ICON 1 resource with 3752 bytes
85246 INFO: Writing RT_ICON 2 resource with 2216 bytes
85247 INFO: Writing RT_ICON 3 resource with 1384 bytes
85247 INFO: Writing RT_ICON 4 resource with 37019 bytes
85247 INFO: Writing RT_ICON 5 resource with 9640 bytes
85247 INFO: Writing RT_ICON 6 resource with 4264 bytes
85247 INFO: Writing RT_ICON 7 resource with 1128 bytes
85258 INFO: Updating manifest in C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\build\Exhaust_fan\run.exe.gd0sz2zk
85259 INFO: Updating resource type 24 name 1 language 0
85262 INFO: Appending archive to EXE C:\Users\lim.jouhui\Desktop\DRMSA\Exploratory analysis\Electrical Component Sizing\Exhaust_fan\dist\Exhaust_fan.exe
85319 INFO: Building EXE from EXE-00.toc completed successfully.

When running exe file, CMD shows:
Exhaust_fan.py:9689: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.
To accept the future behavior, pass 'sort=False'.
To retain the current behavior and silence the warning, pass 'sort=True'.
Exhaust_fan.py:9709: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.
To accept the future behavior, pass 'sort=False'.
To retain the current behavior and silence the warning, pass 'sort=True'.
Exception in Tkinter callback
Traceback (most recent call last):
  File "xlsxwriter\workbook.py", line 320, in close
  File "xlsxwriter\workbook.py", line 638, in _store_workbook
  File "xlsxwriter\workbook.py", line 636, in _store_workbook
  File "zipfile.py", line 1207, in __init__
FileNotFoundError: [Errno 2] No such file or directory: 'Downloads/Exhaust_Fan_Design_Data.xlsx'

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "tkinter\__init__.py", line 1705, in __call__
  File "Exhaust_fan.py", line 9736, in printexcel
  File "pandas\io\excel\_xlsxwriter.py", line 195, in save
  File "xlsxwriter\workbook.py", line 322, in close
xlsxwriter.exceptions.FileCreateError: [Errno 2] No such file or directory: 'Downloads/Exhaust_Fan_Design_Data.xlsx'

**Comment by rokm** [comment, marked as answer]
>  FileNotFoundError: [Errno 2] No such file or directory: 'Downloads/Exhaust_Fan_Design_Data.xlsx'
>  xlsxwriter.exceptions.FileCreateError: [Errno 2] No such file or directory: 'Downloads/Exhaust_Fan_Design_Data.xlsx'

I think the error message is quite self-explanatory - the program cannot find the .xlsx file in the given relative path. Which means that in frozen program, the working directory is not where you think it is (or where it was for the Jupyter Notebook).

So either you need to use absolute file paths (if this is supposed to be an external file), or anchor your relative paths to application's directory (as per [docs](https://pyinstaller.readthedocs.io/en/stable/runtime-information.html#using-file)).

**Reply by jou-hui** [reply]
Yes, the working directory is not where I think, after I revise the working directory. It works. Thanks.
